processor/spanmetricsprocessor/processor.go

```diff
@@ -140,16 +191,215 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
+		mCalls.SetDataType(pdata.MetricDataTypeIntSum)
+		mCalls.SetName("calls")
+		mCalls.IntSum().DataPoints().Append(dpCalls)
+		mCalls.IntSum().SetAggregationTemporality(pdata.AggregationTemporalityCumulative)
+		ilm.Metrics().Append(mCalls)
+	}
+	return nil
 }
 
 // aggregateMetrics aggregates the raw metrics from the input trace data.
 // Each metric is identified by a key that is built from the service name
 // and span metadata such as operation, kind, status_code and any additional
 // dimensions the user has configured.
-func (p *processorImp) aggregateMetrics(traces pdata.Traces) {
-	// TODO: Add implementation
+func (p *processorImp) aggregateMetrics(traces pdata.Traces) error {
+	for i := 0; i < traces.ResourceSpans().Len(); i++ {
+		rspans := traces.ResourceSpans().At(i)
+		r := rspans.Resource()
+
+		attr, ok := r.Attributes().Get(conventions.AttributeServiceName)
+		if !ok {
+			continue
+		}
+		serviceName := attr.StringVal()
+		if err := p.aggregateMetricsForServiceSpans(rspans, serviceName); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *processorImp) aggregateMetricsForServiceSpans(rspans pdata.ResourceSpans, serviceName string) error {
+	ilsSlice := rspans.InstrumentationLibrarySpans()
+	for j := 0; j < ilsSlice.Len(); j++ {
+		ils := ilsSlice.At(j)
+		spans := ils.Spans()
+		for k := 0; k < spans.Len(); k++ {
+			span := spans.At(k)
+			if err := p.aggregateMetricsForSpan(serviceName, span); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (p *processorImp) aggregateMetricsForSpan(serviceName string, span pdata.Span) error {
+	key, err := p.buildKey(serviceName, span)
+	if err != nil {
+		return err
+	}
+
+	p.lock.Lock()
+	p.updateCallMetrics(key)
+	p.lock.Unlock()
+
+	latency := float64(span.EndTime()-span.StartTime()) / float64(time.Millisecond.Nanoseconds())
+
+	// Binary search to find the latency bucket index.
+	index := sort.SearchFloat64s(p.latencyBounds, latency)
+
+	p.lock.Lock()
+	p.updateLatencyMetrics(key, latency, index)
+	p.lock.Unlock()
+	return nil
+}
+
+// updateCallMetrics increments the call count for the given metric key.
+func (p *processorImp) updateCallMetrics(key string) {
+	p.callSum[key]++
+}
+
+// updateLatencyMetrics increments the histogram counts for the given metric key and bucket index.
+func (p *processorImp) updateLatencyMetrics(key string, latency float64, index int) {
+	if _, ok := p.latencyBucketCounts[key]; !ok {
+		p.latencyBucketCounts[key] = make([]uint64, len(p.latencyBounds))
+	}
+	p.latencySum[key] += latency
+	p.latencyCount[key]++
+	p.latencyBucketCounts[key][index]++
+}
+
+// buildKey builds the metric key from the service name and span metadata
+// such as operation, kind, status_code and any additional dimensions the
+// user has configured.
+func (p *processorImp) buildKey(serviceName string, span pdata.Span) (string, error) {
+	d := p.getDimensions(serviceName, span)
+	p.lock.Lock()
+	lastDimension := p.metricsKeyCache.InsertDimensions(d...)
```

Comment:
The purpose of metricsKeyCache is not clear to me. Is it done to avoid serializing to json on every buildKey call? Is there any evidence this is less expensive? Also why json? Can we serialize trivially by concatenating dimensions and values sequentially? Would that be less expensive than json and most importantly less expensive than the trie-based cache?

Reply:
OK, I can see benchmarking of the cache was done, which is good. It is not clear why not use a much simpler serialization instead of json. It may be potentially faster than the cache itself and eliminate the need for the cache.